pkg/urbit/jets/e/de_json.c

```diff
@@ -0,0 +1,238 @@
+/* j/5/de_json.c
+**
+*/
+#include "all.h"
+#include "pdjson.h"
+
+/* custom code for calling parser imported from:
+** https://github.com/skeeto/pdjson
+*/
+
+typedef struct _u3qedj_coll {
+  // collection (list for array, map for object)
+  u3_noun  col;
+  // store context for recursive arrays/objects:
+  //   - pointer to tail for array
+  //   - key for object
+  union {
+    u3_noun* tel;
+    u3_atom  key;
+  };
+} u3qedj_coll;
+
+static u3qedj_coll*
+_push_stack(const u3a_pile *pil_u)
+{
+  u3qedj_coll *res_u = u3a_push(pil_u);
+  u3a_pile_sane(pil_u);
+
+  res_u->col = u3_nul;
+  res_u->key = u3_none;
+
+  return res_u;
+}
+
+static u3qedj_coll*
+_pop_stack(const u3a_pile *pil_u)
+{
+  return u3a_pop(pil_u);
+}
+
+static void
+_close_stack(const u3a_pile *pil_u)
+{
+  while ( c3n == u3a_pile_done(pil_u) ) {
+    u3qedj_coll *tak_u = u3a_peek(pil_u);
+    u3z(tak_u->col);
+    u3a_drop(pil_u);
+  }
```

Comment:
If you have a type tag in `u3qedj_coll`, you might want to dispatch off of that instead of `json_get_context()`. In any case, having that info two places means you have to handle disagreements (or at least assert), which is a pain.

Reply:
I thought that I had come up with a clever solution to this twice, but both times I was wrong. In the end, I added a `c3_b` flag which is enabled when the stack is storing a key. Please take a look - I don't think this solution requires changing the dispatching or any additional assertions.